keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1266 @@
+
+### Mark Job Complete if some indexes failed
+
+The alternative to the proposed [Job completion](#job-completion) strategy.
+
+Allow execution of all indexes, up to `.spec.runPolicy.maxFailedIndexes` of
+failed indexes. Then, mark the Job `Complete` even if some indexes failed.
+The Job is marked `Failed` only if the number of failed indexes exceeds the
+specified `.spec.runPolicy.maxFailedIndexes` limit, in that case, the `reason`
+field could be `FailedIndexes`, and the `message` field would list the failed
+indexes up to a couple of them.
+
+**Reasons for deferring / rejecting**
+
+This approach is less intuitive to the end-users of the API, compared
+to the proposal. In particular, in some cases it would require custom logic in
+the user's controller to determine if the Job is failed.
+
+### Support backoffLimitPerIndex when restartPolicy=OnFailure
+
+We've considered supporting the backoffLimitPerIndex when pod's `restartPolicy=OnFailure`.
+However, in this mode it is Kubelet's responsibility to restart the pod, while
+the maximal number of restarts would be enforced by the Job controller, which
+is racy in its nature. For example, in-between the checks by the Job controller,
+Kubelet execute more restarts than the specified `.spec.backoffLimit`.
+
+**Reasons for deferring / rejecting**
+
+We believe that this feature can be supported well by using the pod-level API,
+started in this KEP:
+[Add a new field maxRestartTimes to podSpec when running into RestartPolicyOnFailure](https://github.com/kubernetes/enhancements/issues/3322).
+
+Once the pod-level API is done, it could be considered to support `.spec.runPolicy.backoffLimitPerIndex`
+when`restartPolicy=OnFailure` in pod's spec. In this case we could set the pod-level
+`maxRestartTimes` field based on the Job-level `.spec.backoffLimit`, leaving the
+responsibility of enforcing the limit to the Kubelet.
+
+### Set backoffLimitPerIndex at the top level
+
+We could set the backoffLimitPerIndex at top-level in order to put it next to
+the backoffLimit field. For example:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimit: 4
+  backoffLimitPerIndex: true
+  ...
+```
+
+**Reasons for deferring / rejecting**
+
+The backoffLimitPerIndex is also conceptually close to maxFailedIndexes.
+We will consider this option based on the community feedback.
+
+### Enforce backoffLimit and backoffLimitPerIndex
+
+We've also considered to support both `backoffLimit` and integer
+`backoffLimitPerIndex`. In that case the Job would be marked failed if the
+global counter of failures `backoffLimit` is reached, as well as if the
+index-scoped limit is hit.
+
+**Reasons for deferring / rejecting**
+
+We don't see a natural use case for this, so it seems an unnecessary
+complication. Also, the default value for `backoffLimit` of 6 might often be
+not enough when `backoffLimitPerIndex` is used, leading to potential mistakes.
```

Comment:
Another reason why this shouldn't be a warning but a hard fail.

Reply:
Still, the use case (as suggested by @deads2k) for setting custom backoffLimit and backoffLimitPerIndex before downgrade applies. Maybe it is not that important though, I'm fine with hard fail in the first iteration. WDYT @deads2k ?